tests/robustness/multiclient_test/multiclient_test.go

```diff
@@ -0,0 +1,44 @@
+// +build darwin,amd64 linux,amd64
+
+package multiclienttest
+
+import (
+	"context"
+	"strconv"
+	"testing"
+
+	"github.com/kopia/kopia/internal/testlogging"
+	"github.com/kopia/kopia/tests/robustness/engine"
+	"github.com/kopia/kopia/tests/robustness/fiofilewriter"
+	"github.com/kopia/kopia/tests/robustness/multiclient_test/framework"
+	"github.com/kopia/kopia/tests/testenv"
+)
+
+func TestExample(t *testing.T) {
+	fileSize := 40 * 1024 * 1024
+	numFiles := 1
+	numClients := 4
+
+	fileWriteOpts := map[string]string{
+		fiofilewriter.MaxDirDepthField:         strconv.Itoa(1),
+		fiofilewriter.MaxFileSizeField:         strconv.Itoa(fileSize),
+		fiofilewriter.MinFileSizeField:         strconv.Itoa(fileSize),
+		fiofilewriter.MaxNumFilesPerWriteField: strconv.Itoa(numFiles),
+		fiofilewriter.MinNumFilesPerWriteField: strconv.Itoa(numFiles),
+	}
+
+	// Define per-client test actions
+	f := func(ctx context.Context) {
+		_, err := eng.ExecAction(ctx, engine.WriteRandomFilesActionKey, fileWriteOpts)
+		testenv.AssertNoError(t, err)
+	}
+
+	// Create or obtain root context
+	ctx := testlogging.Context(t)
+
+	// Obtain a slice of derived contexts, each wrapped with a unique client
+	ctxs := framework.NewClientContexts(ctx, numClients)
+
+	// Run test actions for each client concurrently
+	framework.RunAllAndWait(ctxs, f)
```

Comment:
A thought: There will likely be multiple test functions, all using the same (global) engine. These tests will run serially but each will acquire and use clients, if using the pattern here.  You may want to consider a mechanism whereby the clients can be deleted from the engine when no longer required.

Some possibilities, not necessarily disjoint:
- Pass `numClients` to `framework.RunAllAndWait`, eliminating the need to call `framework.NewClientContexts`. Then the clean-up would be implicit with the call. Perhaps call this variant `framework.RunNAndWait` and leave the original as-is in case there are situations where there is a need for multiple back-to-back calls with different functions. (naming up to you)
- Provide a `cleanup bool` parameter to `framework.RunAllAndWait` to indicate that the current invocation should clean up.
- Provide a more sophisticated test pattern with a test suite setup that handles the client. In this case, though, we'd have to get back an array of clients and not a context because Go idiom frowns on saving a context in a struct.

Reply:
I noticed this when running tests. We do the cleanup at the suite level so clients linger around for longer than they are needed, and are only cleaned up at the very end.

I like the idea of a `framework.RunNAndWait` but we would have to pass the engine (or snapshotter) in to make cleanup possible. And even still, there are a few issues:
- No Cleanup() method on the `robustness.Snapshotter` interface, so we can't call Cleanup here.
- Current Cleanup method on MultiClientSnapshotter also cleans up the server. So the functionality will need to be broken up into a global and per-test cleanup.

Something like this would be nice:

```go
func RunNAndWait(ctx context.Context, numClients int, f func(context.Context), eng *engine.Engine) {
	ctxs := NewClientContexts(ctx, numClients)
	RunAllAndWait(ctxs, f)
	eng.TestRepo.Cleanup()
}
```

Reply:
Make the harness a global. Suggest:
```
func (th *TestHarness) RunN(ctx ..., numClients int, testFn ...) {
   ctx = NewClientContexts(ctx, numClients)
   th.Run(ctx, true, testFn)
}

func (th *TestHarness) Run(ctx ..., cleanup bool, testFn...) { ... }
```